File: apps-microservices/crawler-monitor-backend/src/lib/alerts.js
/**
 * Alerts rules engine.
 *
 * Pure functions: given recent state (jobs, capacity history, replicas
 * history, callbacks count), evaluate the configured rules and return a
 * normalized list of alerts.
 *
 * Thresholds come from env vars (Phase 3 = hardcoded). Phase 4 will allow
 * runtime configuration.
 */

export const DEFAULT_THRESHOLDS = {
  // error_rate_high: failed/completed >= threshold (over last 1h)
  errorRateThreshold:    parseFloat(process.env.ALERT_ERROR_RATE_THRESHOLD || '0.05'),
  errorRateMinJobs:      parseInt(process.env.ALERT_ERROR_RATE_MIN_JOBS || '5', 10),
  // oom_spike: sum of oom_restart_count over 1h >= threshold
  oomSpikeThreshold:     parseInt(process.env.ALERT_OOM_SPIKE_THRESHOLD || '3', 10),
  // replica_high_cpu_sustained: cpu > threshold for >= duration
  replicaHighCpu:        parseFloat(process.env.ALERT_REPLICA_HIGH_CPU || '0.85'),
  replicaHighCpuDurMs:   parseInt(process.env.ALERT_REPLICA_HIGH_CPU_MIN_DURATION || '240000', 10),
  // capacity_full_sustained: full=true continuously for >= duration
  capacityFullDurMs:     parseInt(process.env.ALERT_CAPACITY_FULL_DURATION || '300000', 10),
  // callbacks_failing: count >= min
  callbacksFailedMin:    parseInt(process.env.ALERT_CALLBACKS_FAILED_MIN || '1', 10),
};

const ONE_HOUR_MS = 60 * 60 * 1000;

/* ---------- Individual rules (pure) ---------- */

export function evalErrorRate(jobs, nowMs, thresholds) {
  const cutoff = nowMs - ONE_HOUR_MS;
  const inWindow = jobs.filter(j => {
    const t = Date.parse(j.start_time);
    return Number.isFinite(t) && t >= cutoff;
  });
  const failed   = inWindow.filter(j => (j.status || '').toLowerCase() === 'failed').length;
  const finished = inWindow.filter(j => ['finished', 'archived'].includes((j.status || '').toLowerCase())).length;
  const completed = failed + finished;
  if (completed < thresholds.errorRateMinJobs) return null;
  const rate = failed / completed;
  if (rate < thresholds.errorRateThreshold) return null;
  return {
    id: 'error_rate_high:1h',
    severity: 'warn',
    kind: 'error_rate_high',
    message: `Taux d'erreur ${(rate * 100).toFixed(1)}% sur 1h (${failed}/${completed})`,
    since: null,
    metadata: { rate, failed, completed, window: '1h', threshold: thresholds.errorRateThreshold },
  };
}

export function evalOomSpike(jobs, nowMs, thresholds) {
  const cutoff = nowMs - ONE_HOUR_MS;
  let total = 0;
  for (const j of jobs) {
    const t = Date.parse(j.start_time);
    if (!Number.isFinite(t) || t < cutoff) continue;
    total += j.oom_restart_count || 0;
  }
  if (total < thresholds.oomSpikeThreshold) return null;
  return {
    id: 'oom_spike:1h',
    severity: 'critical',
    kind: 'oom_spike',
    message: `${total} OOM restarts cumulés sur 1h`,
    since: null,
    metadata: { total, window: '1h', threshold: thresholds.oomSpikeThreshold },
  };
}

/**
 * For a single replica: if the last contiguous streak of points where
 * cpu > threshold has duration >= durMs, emit an alert.
 *
 * `points`: chronological array of { ts, cpu } objects.
 */
export function evalReplicaHighCpu(replicaId, points, nowMs, thresholds) {
  if (!points || points.length === 0) return null;
  const last = points[points.length - 1];
  if ((last.cpu || 0) <= thresholds.replicaHighCpu) return null;
  // Walk back to find streak start
  let streakStart = last.ts;
  for (let i = points.length - 2; i >= 0; i--) {
    if ((points[i].cpu || 0) > thresholds.replicaHighCpu) {
      streakStart = points[i].ts;
    } else {
      break;
    }
  }
  const dur = last.ts - streakStart;
  if (dur < thresholds.replicaHighCpuDurMs) return null;
  return {
    id: `replica_high_cpu:${replicaId}`,
    severity: 'warn',
    kind: 'replica_high_cpu_sustained',
    message: `Replica ${replicaId.slice(0, 12)} : CPU > ${(thresholds.replicaHighCpu * 100).toFixed(0)}% depuis ${Math.floor(dur / 60000)} min`,
    since: streakStart,
    metadata: { replicaId, current_cpu: last.cpu, duration_ms: dur, threshold: thresholds.replicaHighCpu },
  };
}

/**
 * Capacity has been continuously full for >= capacityFullDurMs.
 * `points`: chronological [{ts, full}, ...].
 */
export function evalCapacitySaturation(points, nowMs, thresholds) {
  if (!points || points.length === 0) return null;
  const last = points[points.length - 1];
  if (!last.full) return null;
  let streakStart = last.ts;
  for (let i = points.length - 2; i >= 0; i--) {
    if (points[i].full) {
      streakStart = points[i].ts;
    } else {
      break;
    }
  }
  const dur = last.ts - streakStart;
  if (dur < thresholds.capacityFullDurMs) return null;
  return {
    id: 'capacity_full',
    severity: 'critical',
    kind: 'capacity_full_sustained',
    message: `Capacité saturée depuis ${Math.floor(dur / 60000)} min`,
    since: streakStart,
    metadata: { duration_ms: dur, threshold_ms: thresholds.capacityFullDurMs },
  };
}

export function evalCallbacksFailing(failedCallbackCount, thresholds) {
  if (!failedCallbackCount || failedCallbackCount < thresholds.callbacksFailedMin) return null;
  return {
    id: 'callbacks_failing',
    severity: 'critical',
    kind: 'callbacks_failing',
    message: `${failedCallbackCount} callback${failedCallbackCount > 1 ? 's' : ''} en échec à rejouer`,
    since: null,
    metadata: { count: failedCallbackCount },
  };
}

/* ---------- Aggregator ---------- */

/**
 * Evaluate every rule and return the non-null alerts.
 *
 * `inputs`:
 *   { jobs, capacityPoints, replicasHistory: { id: [{ts, cpu}, ...] },
 *     failedCallbackCount }
 */
export function evaluateAlerts(inputs, nowMs = Date.now(), thresholds = DEFAULT_THRESHOLDS) {
  const out = [];

  const er = evalErrorRate(inputs.jobs || [], nowMs, thresholds);
  if (er) out.push(er);

  const oom = evalOomSpike(inputs.jobs || [], nowMs, thresholds);
  if (oom) out.push(oom);

  const cap = evalCapacitySaturation(inputs.capacityPoints || [], nowMs, thresholds);
  if (cap) out.push(cap);

  const cb = evalCallbacksFailing(inputs.failedCallbackCount || 0, thresholds);
  if (cb) out.push(cb);

  for (const [id, points] of Object.entries(inputs.replicasHistory || {})) {
    const a = evalReplicaHighCpu(id, points, nowMs, thresholds);
    if (a) out.push(a);
  }

  // Sort: critical first, then by kind for stability
  const sevWeight = { critical: 0, warn: 1, info: 2 };
  out.sort((a, b) => (sevWeight[a.severity] - sevWeight[b.severity]) || a.kind.localeCompare(b.kind));
  return out;
}

File: apps-microservices/crawler-monitor-backend/internal/domain/alerts/alerts.go
// Package alerts implements the rules engine for the crawler monitor.
//
// Pure functions — given recent state (jobs, capacity history, replicas history,
// failed callback count), evaluate the configured rules and return a normalized
// list of alerts.
//
// Mirrors src/lib/alerts.js exactly (same IDs, severities, metadata shapes, sort order).
package alerts

import (
	"fmt"
	"math"
	"sort"
	"strings"

	"github.com/Hellopro-fr/crawler-monitor-backend/internal/datetime"
)

// Thresholds holds all configurable alert thresholds.
type Thresholds struct {
	ErrorRateThreshold  float64
	ErrorRateMinJobs    int
	OomSpikeThreshold   int
	ReplicaHighCpu      float64
	ReplicaHighCpuDurMs int64
	CapacityFullDurMs   int64
	CallbacksFailedMin  int
}

// DefaultThresholds returns the hardcoded Phase 3 thresholds.
// Mirrors JS DEFAULT_THRESHOLDS (env-resolved values).
func DefaultThresholds() Thresholds {
	return Thresholds{
		ErrorRateThreshold:  0.05,
		ErrorRateMinJobs:    5,
		OomSpikeThreshold:   3,
		ReplicaHighCpu:      0.85,
		ReplicaHighCpuDurMs: 240_000,
		CapacityFullDurMs:   300_000,
		CallbacksFailedMin:  1,
	}
}

const oneHourMs = int64(60 * 60 * 1000)

// Alert is a normalized alert object.
type Alert struct {
	ID       string         `json:"id"`
	Severity string         `json:"severity"`
	Kind     string         `json:"kind"`
	Message  string         `json:"message"`
	Since    *int64         `json:"since"`
	Metadata map[string]any `json:"metadata"`
}

// Job is the minimal job view needed for alert evaluation.
type Job struct {
	StartTime       string
	Status          string
	OomRestartCount int
}

// CapacityPoint is a single capacity history snapshot.
type CapacityPoint struct {
	Ts   int64
	Full bool
}

// CpuPoint is a single CPU history sample for one replica.
type CpuPoint struct {
	Ts  int64
	CPU float64
}

/* ---------- Individual rules ---------- */

// EvalErrorRate evaluates the error_rate_high rule over the last 1h.
// Returns nil if the threshold is not exceeded.
func EvalErrorRate(jobs []Job, nowMs int64, t Thresholds) *Alert {
	cutoff := nowMs - oneHourMs
	var failed, finished int
	for _, j := range jobs {
		ts := parseJobTimeMs(j.StartTime)
		if ts < 0 || ts < cutoff {
			continue
		}
		status := strings.ToLower(j.Status)
		if status == "failed" {
			failed++
		} else if status == "finished" || status == "archived" {
			finished++
		}
	}
	completed := failed + finished
	if completed < t.ErrorRateMinJobs {
		return nil
	}
	rate := float64(failed) / float64(completed)
	if rate < t.ErrorRateThreshold {
		return nil
	}
	return &Alert{
		ID:       "error_rate_high:1h",
		Severity: "warn",
		Kind:     "error_rate_high",
		Message:  fmt.Sprintf("Taux d'erreur %.1f%% sur 1h (%d/%d)", rate*100, failed, completed),
		Since:    nil,
		Metadata: map[string]any{
			"rate":      rate,
			"failed":    failed,
			"completed": completed,
			"window":    "1h",
			"threshold": t.ErrorRateThreshold,
		},
	}
}

// EvalOomSpike evaluates the oom_spike rule: sum of oom_restart_count >= threshold over 1h.
func EvalOomSpike(jobs []Job, nowMs int64, t Thresholds) *Alert {
	cutoff := nowMs - oneHourMs
	var total int
	for _, j := range jobs {
		ts := parseJobTimeMs(j.StartTime)
		if ts < 0 || ts < cutoff {
			continue
		}
		total += j.OomRestartCount
	}
	if total < t.OomSpikeThreshold {
		return nil
	}
	return &Alert{
		ID:       "oom_spike:1h",
		Severity: "critical",
		Kind:     "oom_spike",
		Message:  fmt.Sprintf("%d OOM restarts cumulés sur 1h", total),
		Since:    nil,
		Metadata: map[string]any{
			"total":     total,
			"window":    "1h",
			"threshold": t.OomSpikeThreshold,
		},
	}
}

// EvalReplicaHighCpu evaluates sustained high CPU for a single replica.
// points must be in chronological order (oldest first).
func EvalReplicaHighCpu(replicaID string, points []CpuPoint, nowMs int64, t Thresholds) *Alert {
	if len(points) == 0 {
		return nil
	}
	last := points[len(points)-1]
	if last.CPU <= t.ReplicaHighCpu {
		return nil
	}
	// Walk back to find streak start.
	streakStart := last.Ts
	for i := len(points) - 2; i >= 0; i-- {
		if points[i].CPU > t.ReplicaHighCpu {
			streakStart = points[i].Ts
		} else {
			break
		}
	}
	dur := last.Ts - streakStart
	if dur < t.ReplicaHighCpuDurMs {
		return nil
	}
	// Slice replicaID to 12 chars for the message, matching JS slice(0,12).
	shortID := replicaID
	if len(shortID) > 12 {
		shortID = shortID[:12]
	}
	cpuPct := math.Round(t.ReplicaHighCpu * 100)
	durMin := int(math.Floor(float64(dur) / 60000))
	since := streakStart
	return &Alert{
		ID:       fmt.Sprintf("replica_high_cpu:%s", replicaID),
		Severity: "warn",
		Kind:     "replica_high_cpu_sustained",
		Message:  fmt.Sprintf("Replica %s : CPU > %.0f%% depuis %d min", shortID, cpuPct, durMin),
		Since:    &since,
		Metadata: map[string]any{
			"replicaId":   replicaID,
			"current_cpu": last.CPU,
			"duration_ms": dur,
			"threshold":   t.ReplicaHighCpu,
		},
	}
}

// EvalCapacitySaturation evaluates sustained capacity saturation.
// points must be in chronological order (oldest first).
func EvalCapacitySaturation(points []CapacityPoint, nowMs int64, t Thresholds) *Alert {
	if len(points) == 0 {
		return nil
	}
	last := points[len(points)-1]
	if !last.Full {
		return nil
	}
	streakStart := last.Ts
	for i := len(points) - 2; i >= 0; i-- {
		if points[i].Full {
			streakStart = points[i].Ts
		} else {
			break
		}
	}
	dur := last.Ts - streakStart
	if dur < t.CapacityFullDurMs {
		return nil
	}
	durMin := int(math.Floor(float64(dur) / 60000))
	since := streakStart
	return &Alert{
		ID:       "capacity_full",
		Severity: "critical",
		Kind:     "capacity_full_sustained",
		Message:  fmt.Sprintf("Capacité saturée depuis %d min", durMin),
		Since:    &since,
		Metadata: map[string]any{
			"duration_ms":  dur,
			"threshold_ms": t.CapacityFullDurMs,
		},
	}
}

// EvalCallbacksFailing evaluates whether the failed-callback count meets the minimum.
func EvalCallbacksFailing(failedCallbackCount int, t Thresholds) *Alert {
	if failedCallbackCount < t.CallbacksFailedMin {
		return nil
	}
	suffix := ""
	if failedCallbackCount > 1 {
		suffix = "s"
	}
	return &Alert{
		ID:       "callbacks_failing",
		Severity: "critical",
		Kind:     "callbacks_failing",
		Message:  fmt.Sprintf("%d callback%s en échec à rejouer", failedCallbackCount, suffix),
		Since:    nil,
		Metadata: map[string]any{
			"count": failedCallbackCount,
		},
	}
}

/* ---------- Aggregator ---------- */

// Inputs holds all the data needed by Evaluate.
type Inputs struct {
	Jobs                []Job
	CapacityPoints      []CapacityPoint
	ReplicasHistory     map[string][]CpuPoint // replicaID -> chronological CPU points
	FailedCallbackCount int
}

// Evaluate runs every rule and returns the non-nil alerts, sorted critical-first
// then by kind (lexicographic). Mirrors JS evaluateAlerts.
func Evaluate(inputs Inputs, nowMs int64, t Thresholds) []Alert {
	var out []Alert

	if a := EvalErrorRate(inputs.Jobs, nowMs, t); a != nil {
		out = append(out, *a)
	}
	if a := EvalOomSpike(inputs.Jobs, nowMs, t); a != nil {
		out = append(out, *a)
	}
	if a := EvalCapacitySaturation(inputs.CapacityPoints, nowMs, t); a != nil {
		out = append(out, *a)
	}
	if a := EvalCallbacksFailing(inputs.FailedCallbackCount, t); a != nil {
		out = append(out, *a)
	}
	for id, points := range inputs.ReplicasHistory {
		if a := EvalReplicaHighCpu(id, points, nowMs, t); a != nil {
			out = append(out, *a)
		}
	}

	sevWeight := map[string]int{"critical": 0, "warn": 1, "info": 2}
	sort.SliceStable(out, func(i, j int) bool {
		wi := sevWeight[out[i].Severity]
		wj := sevWeight[out[j].Severity]
		if wi != wj {
			return wi < wj
		}
		return strings.Compare(out[i].Kind, out[j].Kind) < 0
	})

	if out == nil {
		out = []Alert{}
	}
	return out
}

/* ---------- helpers ---------- */

// parseJobTimeMs parses an ISO timestamp string and returns UnixMilli, or -1.
func parseJobTimeMs(s string) int64 {
	return datetime.ParseStringMs(s)
}
